daemon/mgr/image.go

```diff
@@ -161,12 +169,80 @@ func (mgr *ImageManager) GetImage(ctx context.Context, idOrRef string) (*types.I
 }
 
 // ListImages lists images stored by containerd.
-func (mgr *ImageManager) ListImages(ctx context.Context, filter ...string) ([]types.ImageInfo, error) {
-	// TODO: support filter functionality
+func (mgr *ImageManager) ListImages(ctx context.Context, filter filters.Args) ([]types.ImageInfo, error) {
 	ctrdImageInfos := mgr.localStore.ListCtrdImageInfo()
 	imgInfos := make([]types.ImageInfo, 0, len(ctrdImageInfos))
 
+	if err := filter.Validate(acceptedImageFilterTags); err != nil {
+		return nil, err
+	}
+
+	var (
+		beforeFilter, sinceFilter *types.ImageInfo
+		beforeTime, sinceTime     time.Time
+		err                       error
+	)
+
+	beforeImages := filter.Get("before")
+	if len(beforeImages) > 0 {
+		// use the last one if multi before images was assigned
+		beforeFilter, err = mgr.GetImage(ctx, beforeImages[len(beforeImages)-1])
+		if err != nil {
+			return nil, err
+		}
+		beforeTime, err = time.Parse(utils.TimeLayout, beforeFilter.CreatedAt)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	sinceImages := filter.Get("since")
+	if len(sinceImages) > 0 {
+		// use the last one if multi since images was assigned
+		sinceFilter, err = mgr.GetImage(ctx, sinceImages[len(sinceImages)-1])
+		if err != nil {
+			return nil, err
+		}
+		sinceTime, err = time.Parse(utils.TimeLayout, sinceFilter.CreatedAt)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	for _, img := range ctrdImageInfos {
+		if beforeFilter != nil {
+			if img.OCISpec.Created.Equal(beforeTime) || img.OCISpec.Created.After(beforeTime) {
+				continue
+			}
+		}
+		if sinceFilter != nil {
+			if img.OCISpec.Created.Equal(sinceTime) || img.OCISpec.Created.Before(sinceTime) {
+				continue
+			}
+		}
+
+		if filter.Contains("reference") {
+			var found bool
+			for _, ref := range mgr.localStore.GetReferences(img.ID) {
+				for _, pattern := range filter.Get("reference") {
```

Comment:
could we retrieve `filter.Get("reference")` before? I think the `Get` will allocate new memory each loop.

Reply:
seems much condition here

Reply:
what is the condition?